compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala

```diff
@@ -1336,4 +1336,18 @@ object messages {
           |out the parameter list when extending it.
           |"""
   }
+
+  case class VarValParametersMayNotBeCallByName(name: Names.TermName, mutable: Boolean)(implicit ctx: Context)
+    extends Message(VarValParametersMayNotBeCallByNameID) {
+    val msg = s"${if (mutable) "`var'" else "`val'"} parameters may not be call-by-name"
+    val kind = "Syntax"
+    val explanation =
+      hl"""Parameters of classes and traits may no be call-by-name. In case you
+          |want the parameter to be evaluated on demand, consider using a function
+          |and a lazy value in the class such as
+          |  ${s"class MyClass(${name}Tick: () => String) {"}
+          |  ${s"  lazy val $name = ${name}Tick"}
```

Comment:
But that's different from what a by-name would do, the closest thing would be to use a `def`, not a `lazy val`:
```class MyClass(fooFun: () => Int)) {
  def foo = fooFun()
```